components/builder-api-client/src/builder.rs

```diff
@@ -667,29 +692,34 @@ impl BuilderAPIProvider for BuilderAPIClient {
                       token: &str,
                       progress: Option<Self::Progress>)
                       -> Result<()> {
+        debug!("Uploading origin key: {}, {}", origin, revision);
+
         let path = format!("depot/origins/{}/keys/{}", &origin, &revision);
-        let mut file =
+        let file =
             File::open(src_path).map_err(|e| Error::KeyReadError(src_path.to_path_buf(), e))?;
         let file_size = file.metadata()
                             .map_err(|e| Error::KeyReadError(src_path.to_path_buf(), e))?
                             .len();
 
-        let result = if let Some(mut progress) = progress {
+        let resp = if let Some(mut progress) = progress {
             progress.size(file_size);
-            let mut reader = TeeReader::new(file, progress);
+            let reader = TeeReader::new(file, progress);
+            let body = Body::sized(reader, file_size);
             self.add_authz(self.0.post(&path), token)
-                .body(Body::SizedBody(&mut reader, file_size))
-                .send()
+                .body(body)
+                .send()?
         } else {
+            let body = Body::sized(file, file_size);
             self.add_authz(self.0.post(&path), token)
-                .body(Body::SizedBody(&mut file, file_size))
-                .send()
+                .body(body)
+                .send()?
         };
-        match result {
-            Ok(Response { status: StatusCode::Created,
-                          .. }) => Ok(()),
-            Ok(response) => Err(err_from_response(response)),
-            Err(e) => Err(Error::from(e)),
+        debug!("Status: {:?}", resp.status());
+
+        if resp.status() != StatusCode::OK {
+            Err(err_from_response(resp))
+        } else {
+            Ok(())
```

Comment:
You could also add something like this to the trait to start abstracting the common "parse the response as a JSON and return the type" pattern:

```rust
    fn parse_json_if_status<T: serde::de::DeserializeOwned>(mut self,
                                                            code: reqwest::StatusCode)
                                                            -> Result<T> {
        if self.status() == code {
            Ok(self.json()?)
        } else {
            Err(err_from_response(self))
        }
    }
```
then just call

```rust
resp.parse_json_if_status(StatusCode::OK)
```
etc.

Reply:
(these suggestions can all be modified, of course; I just wanted to provide an example)

Reply:
Cool... I like it!

Reply:
Between these two approaches, I'm inclined to go with the extension trait. It solves a couple of different issues - the error handling, as well as abstracting out things like headers in a nicer way. It also lets us return error for specific status codes which the error_for_status does not. Also the From<reqwest::Error> error logic doesn't seem right when moving err_from_response there, since we would have to handle both error from status case as well as a regular reqwest error case.